mr cache: Proposed fix for handling page fault notifications

shefty opened the pull request:
See discussion for PR #5067

shefty commented:
Note: This PR is currently untested, for review only.

shefty commented:
@wckzhang - Can you please review?  If these changes look good to you, I will merge.

wckzhang commented:
Beginning review

wckzhang commented on prov/util/src/util_mr_cache.c:
```diff
@@ -364,6 +358,19 @@ static struct ofi_mr_entry *ofi_mr_rbt_find(struct ofi_mr_storage *storage,
+{
+	struct ofi_rbnode *node;
+
+	node = ofi_rbmap_search(storage->storage, (void *) key,
```
This still does not resolve the issue of adjacent buffers.

Ex. Find overlap will match buffers A and B if:
A: iov_base = 0x0, iov_len = 1
B: iov_base = 0x1, iov_len = 1
(Simplified, but multiply by page size)
If B experiences a page fault, it should not matter to A. We need to exclude the adjacent buffer match here.

shefty replied:
Yes, this will currently match and uncache adjacent buffers incorrectly.  That shouldn't affect correctness, however, only performance.  We just need to introduce a new match function.

wckzhang replied:
Yes, this should work as is. As soon as the new match function is introduced, I will approve this PR.

shefty commented:
Updated - I added a patch that fixes the return value from ofi_iov_end().  That function now returns the address of the last byte of the buffer.  The effect is that util_mr_find_overlap() now only matches if the buffers actually overlap.  I believe this is the more intuitive behavior.  ofi_iov_left() and ofi_iov_right() also return the more natural result.

A side effect of this change is that adjacent, non-overlapping regions will no longer automatically be merged.  For MR cache merging, I think we want a different merge function (e.g. util_mr_find_mergeable()) that could allow merging regions separated by some small amount.  E.g. If two buffers are separated by less than a page, it would make sense to merge them.  I'm leaving this as a future enhancements however.

wckzhang commented:
This will fix the adjacent regions issues. I agree that in the future we should add something that will merge regions < 1 page apart since they will have to be pinned regardless. Alternatively, we could just page align with len +1 in the merge regions function so this will happen automatically.

shefty commented:
Intel CI failure is some weirdness slowdown with shm provider running atomics.
